relay-server/src/utils/envelope_context.rs

```diff
@@ -89,82 +117,39 @@ impl EnvelopeContext {
                 self.summary.profile_quantity,
             );
         }
-    }
 
-    /// Records an outcome scoped to this envelope's context.
-    pub fn track_outcome(&self, outcome: Outcome, category: DataCategory, quantity: usize) {
-        let outcome_aggregator = OutcomeAggregator::from_registry();
-        outcome_aggregator.do_send(TrackOutcome {
-            timestamp: self.received_at,
-            scoping: self.scoping,
-            outcome,
-            event_id: self.event_id,
-            remote_addr: self.remote_addr,
-            category,
-            // Quantities are usually `usize` which lets us go all the way to 64-bit on our
-            // machines, but the protocol and data store can only do 32-bit.
-            quantity: quantity as u32,
-        });
+        self.finish(RelayCounters::EnvelopeRejected);
     }
 
     /// Returns scoping stored in this context.
     pub fn scoping(&self) -> Scoping {
         self.scoping
     }
 
-    /// Returns the event id of this context, if any.
-    pub fn event_id(&self) -> Option<EventId> {
-        self.event_id
+    /// Returns the instant at which the envelope was received at this Relay.
+    ///
+    /// This is the monotonic time equivalent to [`received_at`](Self::received_at).
+    pub fn start_time(&self) -> Instant {
+        self.start_time
     }
 
     /// Returns the time at which the envelope was received at this Relay.
+    ///
+    /// This is the date time equivalent to [`start_time`](Self::start_time).
     pub fn received_at(&self) -> DateTime<Utc> {
         self.received_at
     }
-}
 
-/// Extension trait for [`Addr`] to log [`Outcome`] for failed messages.
-pub trait SendWithOutcome<A> {
-    /// Sends an asynchronous message to the actor, tracking outcomes on failure.
-    ///
-    /// Communication channel to the actor is bounded. If the returned `Future` object get dropped,
-    /// the message cancels.
-    ///
-    /// If the actor rejects the message, an `Invalid` outcome with internal reason is logged. Any
-    /// error within the message result is not tracked by this function.
-    fn send_tracked<M>(
-        &self,
-        message: M,
-        envelope_context: EnvelopeContext,
-    ) -> ResponseFuture<M::Result, MailboxError>
-    where
-        M: Message,
-        A: Handler<M>,
-        A::Context: ToEnvelope<A, M>,
-        M: Message + Send + 'static,
-        M::Result: Send;
+    /// Resets inner state to ensure there's no more logging.
+    fn finish(&mut self, counter: RelayCounters) {
+        relay_statsd::metric!(counter(counter) += 1);
+        relay_statsd::metric!(timer(RelayTimers::EnvelopeTotalTime) = self.start_time.elapsed());
```

Comment:
Note that these metrics are now also logged for envelopes that are rejected in the endpoint (`handle_store_like_request`). This is actually more truthful.